Shopify/dawn
# Try adding global scroll animations to Dawn

### PR Summary: 

This PR adds basic reveal animations to Dawn. 

Here's how it looks: 

https://user-images.githubusercontent.com/1202812/218783665-797dd11d-5044-4ac7-bebf-6b5b1c7a9d79.mp4

### What approach did you take?

The reveal animations are turned on globally via a single checkbox in Theme Settings. The PR uses a light Javascript `IteractionObserver` to add a `scrolled-into-view` class to various elements. This triggers a light CSS animation. 

Some additional notes: 

- The JS was lifted from @stufreen's [stu/scroll-trigger](https://github.com/Shopify/dawn/compare/stu/scroll-trigger) branch.
- It respects [prefers-reduced-motion](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion), and turns itself off if the `no-js` class is present.
- Some elements (cards, for example) are set to load in sequentially. ~~The `nth-child()` logic could use some close 👀 — every now and then the load order isn't 100% left-to-right as expected.~~ (Fixed in https://github.com/Shopify/dawn/pull/2141/commits/514e3f0909f42c6d4a06f8873d6d5b9a2daebf51)   
- I couldn't figure out where exactly to place the "Motion" item sequentially within Theme Settings — the order there feels a bit random

Followup tasks: 

- [ ] In general, some bug hunting. 
- [x] Test with no Javascript
- [ ] In the editor, new sections appear invisible when they're added in. 
- [ ] Test with an eye for potential special cases. (For example, do we want to modify the animation for a specific section?)

### Visual impact on existing themes

None at this point — this is currently built as an opt-in. We should re-visit this though before launch. 

### Testing steps/scenarios

1. Visit [the demo store](https://os2-demo.myshopify.com/admin/themes/137469296662/editor).
2. Visit Theme Settings > Motion
3. Check the "Reveal sections on scroll" checkbox.
4. Scroll around and find bugs! 

### Checklist
